bokeh/model.py

```diff
@@ -712,53 +679,52 @@ def _visit_value_and_its_immediate_references(obj, visitor):
             _visit_value_and_its_immediate_references(key, visitor)
             _visit_value_and_its_immediate_references(value, visitor)
 
-class _ModelInDocument(object):
-    '''
-
-    '''
-
-    def __init__(self, models):
-        # 'models' can be a single Model, a single Document, or a list of either
-        from .document import Document
-
-        self._to_remove_after = []
-        if not isinstance(models, list):
-            models = [models]
 
-        self._doc = _find_some_document(models)
-        if self._doc is None:
-            # oh well - just make up a doc
-            self._doc = Document()
+@contextmanager
+def _ModelInDocument(models):
+    doc = _find_existing_docs(models)
+    models_to_dedoc = _add_doc_to_models(doc, models)
 
-        for model in models:
-            if isinstance(model, Model):
-                if model.document is None:
-                    self._to_remove_after.append(model)
+    yield models
 
-    def __exit__(self, type, value, traceback):
-        for model in self._to_remove_after:
-            model.document.remove_root(model)
+    for model in models_to_dedoc:
+        doc.remove_root(model)
 
-    def __enter__(self):
-        for model in self._to_remove_after:
-            self._doc.add_root(model)
+def _find_existing_docs(models):
+    from bokeh.io import curdoc; curdoc
+    from bokeh.document import Document; Document
 
-@contextmanager
-def _ModelInEmptyDocument(model):
-    '''
+    existing_docs = set(m if isinstance(m, Document) else m.document for m in models)
+    existing_docs.discard(None)
 
-    '''
-    from .document import Document
-    full_doc = _find_some_document([model])
-
-    model._document = None
-    for ref in model.references():
-        ref._document = None
-    empty_doc = Document()
-    empty_doc.add_root(model)
-
-    yield model
+    if len(existing_docs) == 0:
+        # no existing docs, use the current doc
+        doc = curdoc()
```

Comment:
It's one way, but it seems that copying any theme to a new `Document` should also be feasible?